### PR TITLE
Add supplementary groups support for graphene

### PR DESCRIPTION
**Pull request by zhangckid**
We tested memcached in graphene meets one issue:
The integrate tested version can running in graphene
(https://github.com/memcached/memcached/archive/master.zip),
but the latest version can not running in graphene(1.5.10).

It always print like that:
failed to drop supplementary groups

And memcached can't working.

We investigated the latest version of memcached and found that
it used the setgroups syscall lead to this problem. This syscall is not
supported by graphene in wiki page.
In this series, we add two new syscall for graphene:"setgroups" and "getgroups"
to solve this problem.

Signed-off-by: Zhang Chen <chen.zhang@intel.com>

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/252)

### CODE REVIEW

**Comment by zhangckid** [comment]
> Can one of the admins verify this patch?

Yes, Li Lei have reviewed this patch.

File: LibOS/shim/src/sys/shim_getpid.c [hunk 1]
```diff
@@ -88,6 +89,67 @@ int shim_do_setgid (gid_t gid)
...
+    struct shim_thread * cur = get_cur_thread();
+    struct groups_info * groups;
+    int i;
+
+    if (gidsetsize > NGROUPS_MAX)
+        return -EINVAL;
+
+    groups = calloc(1, sizeof(struct groups_info));
+    if (!groups)
+        return -1;
```
Comment (chiache) [inline LibOS/shim/src/sys/shim_getpid.c]: Should return -ENOMEM;

**Comment by donporter** [comment]
Ok to test

File: LibOS/shim/src/sys/shim_getpid.c [hunk 2]
```diff
@@ -88,6 +89,67 @@ int shim_do_setgid (gid_t gid)
...
+    }
+    cur->groups = groups;
+
+    return 0;
+}
+
+int shim_do_getgroups (int gidsetsize, gid_t * grouplist)
+{
+    struct shim_thread * cur = get_cur_thread();
+    int i, j;
```
Comment (chiache) [inline LibOS/shim/src/sys/shim_getpid.c]: I will suggest using a less generic variable name than "i", such as "cur_groups_size".

File: LibOS/shim/src/sys/shim_getpid.c [hunk 3]
```diff
@@ -88,6 +89,67 @@ int shim_do_setgid (gid_t gid)
...
+            if (i > gidsetsize)
+                return -EINVAL;
+
+            for (j = 0; j < i; j++)
+                grouplist[j] = cur->groups->spl_gid[j];
+        }
+    } else {
+        return -1;
+    }
+}
```
Comment (chiache) [inline LibOS/shim/src/sys/shim_getpid.c]: (1) There is no return value at the end of this function.

(2) -1 is returned when the current thread has no supplementary group, while 0 should be returned.

(3) According to the manpage, if "size" is 0, "list" is not modified but the total number of supplementary group IDs for the process is returned.

Comment (zhangckid) [reply LibOS/shim/src/sys/shim_getpid.c]: Good catch! I will address your comments in next version.

Thanks

**Comment by zhangckid** [comment]
Hi~ Tsai and Don,

I have addressed yours comments and rebased this patch.
Please review it again.

Thanks
Zhang Chen

**Comment by donporter** [comment]
Just curious - why did you close the PR?

**Comment by zhangckid** [comment]
Hi Tsai,

I have fixed this issue and updated my branch,
Please review.

Thanks
Zhang Chen

**Comment by zhangckid** [comment]
Hi, 
Have any new for this PR?

Thanks
Zhang Chen

**Comment by zhangckid** [comment]
Hi Don,

I have updated as your comments~
Please review again.

Thanks
Zhang Chen

**Comment by donporter** [comment]
I would like to get this PR landed.  @zhangckid would you mind rebasing this to master again?  You can drop @chiache 's merge master so that it rebases cleanly.

Let's see if this passes CI; if not, I'll take a look at what is failing and get back to you.

**Comment by zhangckid** [comment]
> I would like to get this PR landed. @zhangckid would you mind rebasing this to master again? You can drop @chiache 's merge master so that it rebases cleanly.
> 
> Let's see if this passes CI; if not, I'll take a look at what is failing and get back to you.

Sorry, I didn't get the message from github. I will start to look at it.

**Comment by zhangckid** [comment]
Hi @donporter ,

I'm very confused about use the LTP tests several days, even if I use master branch graphene to run the LTP still got a lot of failures like you said(but I checkout to this branch will fix some getgroups/setgroups cases and caused some result changed). I checked all my environment but nothing found. I have tested the graphene-SGX for LTP, all cases will failed, is it normal? or something I missed?

**Comment by donporter** [comment]
Many of the LTP tests don't work.  Our script (syscalls.sh) just checks that the tests that we think work are still working.  In other words, as we debug them, we add them to the PASSED list, and use Jenkins to catch when a change breaks something that was working.

So the part to pay attention to is at the end, where it says the difference.  As I mentioned, access03 works for me on master, but not your branch.

Also note that we only run these tests on a Linux (non-SGX) PAL.

Let me know if you are still having difficulties reproducing this.

**Comment by zhangckid** [comment]
Sorry for missed this thread.
I have updated codes, please re-check with LTP.
@donporter 

Thanks

**Pull request merged**